### PR TITLE
Feature/room drawer

### PR DESCRIPTION
**Pull request by EliasPaulinoAndrade**
Tasks:
Mockar a casa
Arrastar Móveis

### CODE REVIEW

File: FinalChallenge/House Representation/Models/HouseTiles/WallEndTile.swift [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import SpriteKit
+
+class WallEndTile: HouseTileProtocol {
+    
+    var isFloor = false
+    
+    lazy var description = self.aligntToLeft ? "▶️" : "🔼"
+    
+    let aligntToLeft: Bool
+
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Models/HouseTiles/WallEndTile.swift]: remove spaces between variables

File: FinalChallenge/House Representation/Models/HouseTiles/WallTile.swift [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        let height = (width * floorTexture.size().height) / floorTexture.size().width
+        
+        let definition = SKTileDefinition(texture: floorTexture, size: CGSize(width: width, height: height))
+        
+        if !aligntToLeft {
+            definition.flipHorizontally = true
+        }
+        
+        return definition
+    }()
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Models/HouseTiles/WallTile.swift]: remove spaces between variables

File: FinalChallenge/House Representation/Models/IsometricHouse.swift [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    var groupsGrid: [[SKTileGroup]]
+    
+    var possibleDefinitions: [String: HouseTileProtocol]
+    var possibleGroups: [String: SKTileGroup]
+    
+    var groups: [SKTileGroup]
+    var areas: [FloorArea]
+    
+    var furnitureGrid: [[FurnitureGridItem]]
+
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Models/IsometricHouse.swift]: remove spaces between variables

File: FinalChallenge/House Representation/Models/HouseTiles/FloorTile.swift [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+class FloorTile: HouseTileProtocol {
+    var description = "🈚️"
+    
+    lazy var definition: SKTileDefinition? = {
+        let floorTexture = SKTexture(image: UIImage(named: "realFloor")!)
+        
+        let width: CGFloat = 135
+        let height = (width * floorTexture.size().height) / floorTexture.size().width
+        
+        return SKTileDefinition(texture: floorTexture, size: CGSize(width: width, height: height))
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Models/HouseTiles/FloorTile.swift]: remove spaces between lines

File: FinalChallenge/House Representation/Models/HouseTiles/EmptyTile.swift [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+import SpriteKit
+
+class EmptyTile: HouseTileProtocol {
+    
+    var isFloor = false
+    
+    var definition: SKTileDefinition?
+    
+    var description: String = "⬛️"
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Models/HouseTiles/EmptyTile.swift]: remove spaces between variables

File: FinalChallenge/House Representation/Models/HouseTiles/CornerTile.swift [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    var description = "⤵️"
+    
+    lazy var definition: SKTileDefinition? = {
+        let floorTexture = SKTexture(image: UIImage(named: "realWallCorner")!)
+        
+        let width: CGFloat = 135
+        let height = (width * floorTexture.size().height) / floorTexture.size().width
+                
+        return SKTileDefinition(texture: floorTexture, size: CGSize(width: width, height: height))
+    }()
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Models/HouseTiles/CornerTile.swift]: remove spaces between lines

File: FinalChallenge/House Representation/Models/HouseTiles/CornerSideTile.swift [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        let width: CGFloat = 135
+        let height = (width * floorTexture.size().height) / floorTexture.size().width
+        
+        let definition = SKTileDefinition(texture: floorTexture, size: CGSize(width: width, height: height))
+        
+        if !aligntToLeft {
+            definition.flipHorizontally = true
+        }
+        
+        return definition
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Models/HouseTiles/CornerSideTile.swift]: Remove spaces between lines

File: FinalChallenge/House Representation/Models/Furni Event Handler/FurniEventHandler.swift [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+            return
+        }
+
+        furni.anchorPoint.y = furni.pivotRelativePoint.y
+        furni.anchorPoint.x = furni.pivotRelativePoint.x
+
+        furni.move(toPoint: furniPosition, animated: true)
+        
+        furniParent.furniDidMove(furni, to: furniGridPosition)
+    }
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Models/Furni Event Handler/FurniEventHandler.swift]: Remove spaces between lines

File: FinalChallenge/House Representation/Models/Furni Event Handler/FurniEventHandler.swift [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+              let location = touch.location(in: furni.furniParent),
+              let furniParent = furni.furniParent else {
+            return
+        }
+        
+        let furniGridPosition = furniParent.gridPositionFor(position: location)
+        
+        self.needMove(toPosition: furniGridPosition)
+
+        self.delegate?.showSelectionView(atPosition: furni.position)
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Models/Furni Event Handler/FurniEventHandler.swift]: Remove spaces between lines

File: FinalChallenge/House Representation/Models/Furni Event Handler/FurniEventHandler.swift [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+class FurniEventHandler: MovableEventHandler {
+    
+    var movable: FurniProtocol
+    
+    var mode = FurniMode.normal
+    
+    weak var delegate: FurniEventHandlerDelegate?
+
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Models/Furni Event Handler/FurniEventHandler.swift]: Remove spaces between variables

File: FinalChallenge/House Representation/Models/FloorArea.swift [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    let heightRange: ClosedRange<Int>
+    let widthRange: ClosedRange<Int>
+    
+    var minPosition: GridPosition? {
+        guard let minHeight = heightRange.min(),
+              let minWidth = widthRange.min() else {
+            return nil
+        }
+        
+        return (col: minWidth, row: minHeight)
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Models/FloorArea.swift]: Remove spaces between lines

File: FinalChallenge/House Representation/Models/FloorArea.swift [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    }
+    
+    var maxPosition: GridPosition? {
+        guard let maxHeight = heightRange.max(),
+              let maxWidth = widthRange.max() else {
+            return nil
+        }
+        
+        return (col: maxWidth, row: maxHeight)
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Models/FloorArea.swift]: Remove spaces between lines

File: FinalChallenge/House Representation/Extensions/UITouch+locationInFurniParent.swift [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    func location(in parent: FurniParentProtocol?) -> CGPoint? {
+        if let view = parent as? UIView {
+            return self.location(in: view)
+        }
+        
+        if let node = parent as? SKNode {
+            return self.location(in: node)
+        }
+        
+        return nil
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Extensions/UITouch+locationInFurniParent.swift]: Remove spaces between lines

File: FinalChallenge/House Representation/Extensions/UITouch+locationInFurniParent.swift [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+    func previousLocation(in movable: FurniParentProtocol?) -> CGPoint? {
+        if let view = movable as? UIView {
+            return self.previousLocation(in: view)
+        }
+        
+        if let node = movable as? SKNode {
+            return self.previousLocation(in: node)
+        }
+        
+        return nil
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Extensions/UITouch+locationInFurniParent.swift]: Remove spaces between lines

File: FinalChallenge/House Representation/Nodes/FurniNode/FurniNode.swift [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+    }
+    
+    override var description: String {
+        return "◻️"
+    }
+    
+    var furniParent: FurniParentProtocol?
+
+    var pivotRelativePoint: CGPoint
+
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Nodes/FurniNode/FurniNode.swift]: Remove spaces between lines

File: FinalChallenge/House Representation/Nodes/HomeScene.swift [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+        
+        return cameraNode
+    }()
+    
+    lazy var mapNode: HouseTileMapNode = {
+        let tileSet = SKTileSet(tileGroups: viewModel.house.groups, tileSetType: .isometric)
+        
+        let rows = viewModel.house.grid.count, cols = viewModel.house.grid.first?.count ?? 0
+        
+        let tileMapViewModel = HouseTileMapNodeViewModel(house: viewModel.house)
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Nodes/HomeScene.swift]: Remove spaces between lines

File: FinalChallenge/House Representation/Nodes/HomeScene.swift [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+    
+    @objc func pinchGestureAction(_ sender: UIPinchGestureRecognizer) {
+        if sender.state == .began {
+            previousCameraScale = cameraNode.xScale
+        }
+        cameraNode.setScale(previousCameraScale / sender.scale)
+    }
+    
+    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
+
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Nodes/HomeScene.swift]: Remove space

File: FinalChallenge/House Representation/View Models/FurniNodeViewModel.swift [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+class FurniNodeViewModel: FurniNodeViewModelProtocol {
+    func sizeForTexture(_ texture: SKTexture, withIdealWidth idealWidth: CGFloat) -> CGSize {
+        let ratio = texture.size().width / idealWidth
+        let idealHeight = texture.size().height / ratio
+        
+        return CGSize(width: idealWidth, height: idealHeight)
+    }
+    
+    func zPosition(forGridPosition gridPosition: GridPosition, andGridSize gridSize: GridSize) -> CGFloat {
+
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/View Models/FurniNodeViewModel.swift]: Remove space

File: FinalChallenge/House Representation/View Models/HomeSceneViewModel.swift [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+
+    func cameraPropertiesForZoomInArea(
+        _ area: (min: CGPoint, max: CGPoint),
+        andMaxWidth maxWidth: CGFloat) -> (scale: CGFloat, position: CGPoint)
+    func addFurniture(furniture: FurniProtocol)
+    func deselectAllFurnitures()
+}
+
+class HomeSceneViewModel: HomeSceneViewModelProtocol {
+
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/View Models/HomeSceneViewModel.swift]: Remove space

File: FinalChallenge/House Representation/View Models/HouseTileMapNodeViewModel.swift [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+    
+    func canMoveFurni(_ furni: FurniProtocol, to position: GridPosition) -> Bool
+    func canFlipFurni(_ furni: FurniProtocol) -> Bool
+    func zoomArea(atPosition position: GridPosition, gridPositioningMethod: (GridPosition) -> CGPoint) ->  (min: CGPoint, max: CGPoint)?
+    func furniDidMove(_ furni: FurniProtocol, toPosition position: GridPosition)
+    func furniDidRotate(_ furni: FurniProtocol)
+}
+
+class HouseTileMapNodeViewModel: HouseTileMapNodeViewModelProtocol {
+
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/View Models/HouseTileMapNodeViewModel.swift]: Remove space

File: FinalChallenge/House Representation/Views/HomeView/HomeView.swift [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+//
+//  Created by Elias Paulino on 13/09/19.
+//  Copyright © 2019 Elias Paulino. All rights reserved.
+//
+
+import UIKit
+import SpriteKit
+
+class HomeView: SKView {
+
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Views/HomeView/HomeView.swift]: Remove spaces between lines

File: FinalChallenge/Views/BluePrintView.swift [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+import Foundation
+import UIKit
+
+class BluePrintDrawer {
+    private var drawers: [RoomDrawer] = []
+    
+    private weak var bluePrintView: UIView?
+    
+    private var selected: RoomDrawer?
+
```
Comment (jessicalewinter) [inline FinalChallenge/Views/BluePrintView.swift]: Remove space between variables

File: FinalChallenge/Views/BluePrintView.swift [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+    }
+    
+    func roomAtPosition(position: CGPoint) -> RoomDrawer? {
+        return drawers.filter { (drawer) -> Bool in
+            drawer.contains(point: position)
+        }.first
+    }
+    
+    @objc func viewWasPanned(recognizer: UIPanGestureRecognizer) {
+
```
Comment (jessicalewinter) [inline FinalChallenge/Views/BluePrintView.swift]: remove space

File: FinalChallenge/Views/BluePrintView.swift [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+            currentDrawer?.room.space = CGRect(origin: originPosition, end: panLocation)
+        default:
+            currentDrawer = nil
+            originPosition = nil
+        }
+    }
+}
+
+class BluePrintView: UIView {
+
```
Comment (jessicalewinter) [inline FinalChallenge/Views/BluePrintView.swift]: remove space

File: FinalChallenge/Views/RoomDrawer.swift [hunk 25]
```diff
@@ -0,0 +1,232 @@
...
+import Foundation
+import UIKit
+
+enum CornerType: Int {
+    case topLeft, topRight, bottomRight, bottomLeft
+}
+
+class RoomView: UIView {
+    
+}
```
Comment (jessicalewinter) [inline FinalChallenge/Views/RoomDrawer.swift]: If you don't use this class you should delete it or commented it.

File: FinalChallenge/Views/RoomDrawer.swift [hunk 26]
```diff
@@ -0,0 +1,232 @@
...
+
+class RoomDrawer {
+    var parentView: UIView
+    
+    var stylist: RoomDrawerStylist? = DefaultStylist()
+    
+    var startLocation: CGPoint?
+
```
Comment (jessicalewinter) [inline FinalChallenge/Views/RoomDrawer.swift]: remove space between variables

File: FinalChallenge/House Representation/Nodes/House Tile Node/HouseTileMapNode.swift [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+    func nearPositionFor(position: CGPoint) -> CGPoint {
+        return positionFor(gridPosition: gridPositionFor(position: position))
+    }
+    
+    func positionFor(gridPosition: GridPosition) -> CGPoint {
+        return self.centerOfTile(atColumn: gridPosition.col, row: gridPosition.row)
+    }
+    
+    func canMoveFurni(_ furni: FurniProtocol, to position: GridPosition) -> Bool {
+        viewModel.canMoveFurni(furni, to: position)
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Nodes/House Tile Node/HouseTileMapNode.swift]: add return on this function

File: FinalChallenge/House Representation/Nodes/House Tile Node/HouseTileMapNode.swift [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+    }
+    
+    func canFlipFurni(_ furni: FurniProtocol) -> Bool {
+        viewModel.canFlipFurni(furni)
```
Comment (jessicalewinter) [inline FinalChallenge/House Representation/Nodes/House Tile Node/HouseTileMapNode.swift]: add return on this function

File: FinalChallenge/House Representation/Models/Furni Event Handler/FurniEventHandler.swift (hunk 8 above)
Comment (EliasPaulinoAndrade) [reply FinalChallenge/House Representation/Models/Furni Event Handler/FurniEventHandler.swift]: This are different types of instructions. I think the code is most readable using the spaces.

File: FinalChallenge/House Representation/Models/Furni Event Handler/FurniEventHandler.swift (hunk 9 above)
Comment (EliasPaulinoAndrade) [reply FinalChallenge/House Representation/Models/Furni Event Handler/FurniEventHandler.swift]: This are different types of instructions. I think the code is most readable using the spaces.

File: FinalChallenge/House Representation/Extensions/UITouch+locationInFurniParent.swift (hunk 13 above)
Comment (EliasPaulinoAndrade) [reply FinalChallenge/House Representation/Extensions/UITouch+locationInFurniParent.swift]: This are different types of instructions. I think the code is most readable using the spaces.

**Pull request merged**